mm2src/coins/nft/storage/wasm/wasm_storage.rs

```diff
@@ -521,6 +587,69 @@ impl NftTransferHistoryStorageOps for IndexedDbNftStorage {
         }
         Ok(res.into_iter().collect())
     }
+
+    async fn get_transfers_by_token_address(
+        &self,
+        chain: &Chain,
+        token_address: String,
+    ) -> MmResult<Vec<NftTransferHistory>, Self::Error> {
+        let locked_db = self.lock_db().await?;
+        let db_transaction = locked_db.get_inner().transaction().await?;
+        let table = db_transaction.table::<NftTransferHistoryTable>().await?;
+
+        let index_keys = MultiIndex::new(NftTransferHistoryTable::CHAIN_TOKEN_ADD_INDEX)
+            .with_value(chain.to_string())?
+            .with_value(&token_address)?;
+
+        table
+            .get_items_by_multi_index(index_keys)
+            .await?
+            .into_iter()
+            .map(|(_item_id, item)| transfer_details_from_item(item))
+            .collect()
+    }
+
+    async fn update_transfer_spam_by_token_address(
+        &self,
+        chain: &Chain,
+        token_address: String,
+        possible_spam: bool,
+    ) -> MmResult<(), Self::Error> {
+        let transfers: Vec<NftTransferHistory> = self
+            .get_transfers_by_token_address(chain, token_address.clone())
+            .await?;
+        let locked_db = self.lock_db().await?;
+        let db_transaction = locked_db.get_inner().transaction().await?;
+        let table = db_transaction.table::<NftTransferHistoryTable>().await?;
+
+        for mut transfer in transfers {
+            transfer.common.possible_spam = possible_spam;
+            drop_mutability!(transfer);
+
+            let index_keys = MultiIndex::new(NftTransferHistoryTable::CHAIN_TX_HASH_LOG_INDEX_INDEX)
+                .with_value(chain.to_string())?
+                .with_value(&transfer.common.transaction_hash)?
+                .with_value(transfer.common.log_index)?;
+
+            let item = NftTransferHistoryTable::from_transfer_history(&transfer)?;
+            table.replace_item_by_unique_multi_index(index_keys, &item).await?;
+        }
+        Ok(())
+    }
+
+    async fn get_token_addresses(&self, chain: &Chain) -> MmResult<Vec<Address>, Self::Error> {
+        let locked_db = self.lock_db().await?;
+        let db_transaction = locked_db.get_inner().transaction().await?;
+        let table = db_transaction.table::<NftTransferHistoryTable>().await?;
+
+        let items = table.get_items("chain", chain.to_string()).await?;
+        let mut token_addresses = HashSet::new();
+        for (_item_id, item) in items.into_iter() {
+            let transfer = transfer_details_from_item(item)?;
+            token_addresses.insert(transfer.common.token_address);
+        }
+        Ok(token_addresses.into_iter().collect())
```

Comment:
Well, bcz in sql we can use DISTINCT token addresses and just collect them in vector. but yeah, we can return HashSet in `get_token_addresses`. it would be better.

Reply:
You can indeed get uniquness from SQL using distinct,  or dedup the vector. Current aproach seems unnecessary overhead from any perspective.

Reply:
left HashSet